File: pages/transfer_page.py
from playwright.sync_api import Page

from pages.base_page import BasePage


class TransferPage(BasePage):
    """Transfer Funds — moves money between the customer's own accounts.

    The page is a small JS app: the form posts via XHR and toggles one of
    three panels — #showForm, #showResult ("Transfer Complete!"), #showError.
    """

    URL = "/parabank/transfer.htm"

    FROM_ACCOUNT_SELECT = "#fromAccountId"
    TO_ACCOUNT_SELECT = "#toAccountId"
    AMOUNT_INPUT = "#amount"
    TRANSFER_BUTTON = 'input[value="Transfer"]'
    FORM_PANEL = "#showForm"
    RESULT_PANEL = "#showResult"
    ERROR_PANEL = "#showError"
    AMOUNT_VALIDATION_ERROR = "p[id='amount.errors']"

    def __init__(self, page: Page, base_url: str) -> None:
        super().__init__(page, base_url)

    def open(self) -> "TransferPage":
        self.navigate(self.URL)
        # Account dropdowns are populated by an XHR after page load.
        self.page.locator(f"{self.FROM_ACCOUNT_SELECT} option").first.wait_for(state="attached")
        return self

    def is_on_transfer_page(self) -> bool:
        return "transfer.htm" in self.page.url

    def transfer(self, amount: str, from_index: int = 0, to_index: int = 1) -> None:
        self.page.locator(self.FROM_ACCOUNT_SELECT).select_option(index=from_index)
        self.page.locator(self.TO_ACCOUNT_SELECT).select_option(index=to_index)
        self.fill(self.AMOUNT_INPUT, amount, "amount")
        self.click(self.TRANSFER_BUTTON, "Transfer button")
        # Both outcomes (result or error panel) hide the form.
        self.page.locator(self.FORM_PANEL).wait_for(state="hidden")

    def is_transfer_complete(self) -> bool:
        return self.page.locator(self.RESULT_PANEL).is_visible()

    def has_error(self) -> bool:
        return self.page.locator(self.ERROR_PANEL).is_visible()

    def has_amount_validation_error(self) -> bool:
        errors = self.page.locator(self.AMOUNT_VALIDATION_ERROR)
        return any(errors.nth(i).is_visible() for i in range(errors.count()))

    def available_from_accounts(self) -> int:
        return self.page.locator(f"{self.FROM_ACCOUNT_SELECT} option").count()

File: dotnet/ParabankQa.Tests/Pages/TransferPage.cs
using Allure.Net.Commons;
using Microsoft.Playwright;

namespace ParabankQa.Tests.Pages;

/// <summary>
/// Transfer Funds — a small JS app: the form posts via XHR and toggles one of
/// #showForm / #showResult ("Transfer Complete!") / #showError.
/// </summary>
public sealed class TransferPage(IPage page, string baseUrl) : BasePage(page, baseUrl)
{
    public const string Url = "/parabank/transfer.htm";

    private const string FromAccountSelect = "#fromAccountId";
    private const string ToAccountSelect = "#toAccountId";
    private const string AmountInput = "#amount";
    private const string TransferButton = "input[value=\"Transfer\"]";
    private const string FormPanel = "#showForm";
    private const string ResultPanel = "#showResult";
    private const string ErrorPanel = "#showError";

    public async Task<TransferPage> OpenAsync()
    {
        await NavigateAsync(Url);
        // Account dropdowns are populated by an XHR after page load.
        await Page.Locator($"{FromAccountSelect} option").First.WaitForAsync(
            new() { State = WaitForSelectorState.Attached });
        return this;
    }

    public bool IsOnTransferPage() => Page.Url.Contains("transfer.htm");

    public Task<int> AvailableFromAccountsAsync() =>
        Page.Locator($"{FromAccountSelect} option").CountAsync();

    public Task TransferAsync(string amount) =>
        AllureApi.Step($"Transfer ${amount} between own accounts", async () =>
        {
            await Page.Locator(FromAccountSelect).SelectOptionAsync(new SelectOptionValue { Index = 0 });
            await Page.Locator(ToAccountSelect).SelectOptionAsync(new SelectOptionValue { Index = 1 });
            await FillAsync(AmountInput, amount, "amount");
            await ClickAsync(TransferButton, "Transfer button");
            // Both outcomes (result or error panel) hide the form.
            await Page.Locator(FormPanel).WaitForAsync(new() { State = WaitForSelectorState.Hidden });
        });

    public Task<bool> IsTransferCompleteAsync() => Page.Locator(ResultPanel).IsVisibleAsync();

    public Task<bool> HasErrorAsync() => Page.Locator(ErrorPanel).IsVisibleAsync();
}
